File: OrderItem.cpp
#include "OrderItem.h"

OrderItem::OrderItem(const std::string& name, int64_t price, int quantity)
    : Item(name, price), quantity(quantity) {
    updateSubtotal();
}

OrderItem::OrderItem(const std::string& name, int64_t price, 
                     const std::string& icon_png, int quantity)
    : Item(name, price, icon_png), quantity(quantity), 
      subtotal(price * quantity) {}

int64_t OrderItem::getQuantity() const {
    return quantity;
}

int64_t OrderItem::getSubtotal() const {
    return subtotal;
}

void OrderItem::setPrice(int64_t price) {
    Item::setPrice(price);
    updateSubtotal();
}

void OrderItem::setQuantity(int64_t quantity) {
    this->quantity = quantity;
    updateSubtotal();
}

void OrderItem::updateSubtotal() {
    subtotal = getPrice() * getQuantity();
}

File: src/order_item.rs
use crate::item::Item;
use mongodb::bson::{doc, Document};

#[derive(Debug)]
pub struct OrderItem {
    pub item: Item,
    pub quantity: i64,
    pub subtotal: i64,
}

impl OrderItem {
    pub fn new(name: String, price: i64, quantity: i64) -> Self {
        let item = Item::new(name, price);
        let subtotal = price * quantity;
        Self {
            item,
            quantity,
            subtotal,
        }
    }

    pub fn with_icon(name: String, price: i64, icon_png: String, quantity: i64) -> Self {
        let item = Item::with_icon(name, price, icon_png);
        let subtotal = price * quantity;
        Self {
            item,
            quantity,
            subtotal,
        }
    }

    pub fn update_subtotal(&mut self) {
        self.subtotal = self.item.price * self.quantity;
    }

    pub fn to_document(&self) -> Document {
        doc! {
            "item": {
                "name": &self.item.name,
                "price": self.item.price,
                "icon_png": self.item.icon_png.clone(),
            },
            "quantity": self.quantity,
            "subtotal": self.subtotal,
        }
    }
}
